### ISSUE TITLE
`RollingQuantile` performance with large windows

### ISSUE THREAD

**Issue by fox-ds** [online-ml/river#906]
## Versions
**`river` version**: 0.10.1 
**Python version**: 3.10.2

## Describe your task
Keeping track of a rolling quantile over a large window. Once the RollingQuantile reaches the windows size, it removes the oldest value and appends the new value.  For 'small' windows ( ~ < 100k ) this is no problem and the code runs really fast, but with larger window sizes (say, 500k) once the window is full, there is a significant performance drop when updating the `RollingQuantile`. I think the issue is primarily in the underlying `utils.SortedWindow`.

Furthermore, though the `RollingQuantile` is indeed a streaming algorithm, but there may be some downsides on the memory and computational complexity side when handling large windows (`O(n)`). Because the entire window is kept in memory and being used to calculate the quantile, it's really accurate. However, maybe we could trade in some of this accuracy for performance, e.g. by estimating the quantile just like the regular `Quantile` class, but then windowed (although I currently don't know of a specific algorithm that solves this issue, there are a few papers out there though).

## What kind of performance are you expecting?
Until the window is full: 347195 iterations/s
Once the window is full: **279** iterations/s

## Steps/code to reproduce

```python
from river.stats import RollingQuantile
from tqdm import tqdm

q = 0.997

rolling_q = RollingQuantile(window_size=500_000, q=q)

for x in tqdm(range(500_000)):
    rolling_q.update(x)

for x in tqdm(range(500_000)):
    rolling_q.update(x)

```

**Comment by smastelini** [comment]
Hi @fox-ds, this is indeed problematic from a computational point of view. However, there is not much we can do currently.

IMHO, the main problem is that although multiple quantile estimators enable adding new points, removing old data is problematic. For instance, the P2 algorithm, implemented under `stats.Quantile`, does not allow "reverting". I am not sure if it is possible to modify the algorithm to enable that, nor if other more robust solutions (e.g., t-digest) support such operation.

In the end, `SortedWindow` does what the name implies. So every new element addition/removal implies re-sorting the elements in the window. I would love to see a "reversible" incremental estimator of quantiles, and I have to admit that I am totally noob in that matter. Maybe this solution is out there and it is simple.

**Comment by smastelini** [comment]
Hey @AdilZouitine, would it be possible to keep a full-size binary tree in memory, even if the window is not full? Then instead of removing elements, we could mark them for removal. The tree would need to be rebalanced/rebuilt from time to time, which might become a problem. 

I am not so sure about the rebuild/rebalancing frequency and whether or not this would become an extra parameter (which is not a good thing, IMHO), but we could think about something like that for the future.

**Comment by fox-ds** [comment]
Thanks for the suggestions all, the x3 speed boost already helps a lot! @smastelini @AdilZouitine I think that indeed keeping a binary tree in memory would be the ideal solution, perhaps we could use the `heapq` library for designing such a class. I found [this blogpost ](https://thewayofnada.medium.com/a-very-thorough-solution-to-sliding-window-median-and-some-heap-magics-5091a3ed1cdc) that tries to achieve something similar (be it for the median only), but maybe we could use that for inspiration. I've also started looking into this issue myself, maybe will come up with a solution later. Again, thanks a lot for the help!

**Comment by MaxHalford** [comment]
I found [this](https://aakinshin.net/posts/partitioning-heaps-quantile-estimator/) promising blog post. But it looks complex, so someone needs to dive into it.

**Comment by fox-ds** [comment]
Thanks for the link @MaxHalford. The past week I worked on what at first glance looks like a similar solution - using two heaps that are 'balanced' around the quantile that you want to estimate. I managed to get it up and running really efficiently with `window_size=500_000`, however (in my implementation) the memory usage was still quite high. So I'll have a look at this blog post as well.

**Comment by fox-ds** [comment]
Not yet, but I want to continue with it this week and probably share some code/figures once I'm satisfied with the results.

**Comment by smastelini** [comment]
Hey @fox-ds and @MaxHalford, I've read both blog posts and I really like those ideias! The more I learn, the more I realize I know nothing about anything.

The reference C# implementation @MaxHalford sent also generalizes to arbitrary quantiles :D

**Comment by fox-ds** [comment]
I did continue working on this a few weeks ago, but still need some more time to finalize the results

**Comment by MaxHalford** [comment]
We've just released version 0.13 which includes faster implementations of quantile algorithms, thanks to @AdilZouitine. I think we can now close this issue.

**Comment by smastelini** [comment]
Nice!!!
